linkerd/app/core/src/control.rs

```diff
@@ -145,160 +145,376 @@ pub mod add_origin {
-
-                    let future = self.dns.resolve_one_ip(na.name());
-                    State::Resolve(future, Some((self.inner.clone(), target.clone())))
-                }
-            };
-
-            Init { state }
+        for old_target in old_targets.iter() {
+            if !new_targets.contains(old_target) {
+                removes.push(old_target.addr)
+            }
         }
-    }
 
-    // === impl Init ===
+        (adds, removes)
+    }
 
-    impl<M> Future for Init<M>
+    pub fn resolution_stream<R>(
+        dns: R,
+        address: ControlAddr,
+    ) -> impl Stream<Item = Result<Update<Target>, Error>>
     where
-        M: tower::Service<client::Target>,
+        R: dns::Resolver,
     {
-        type Output = Result<M::Response, Error<M::Error>>;
-
-        fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
-            let mut this = self.project();
-            loop {
-                match this.state.as_mut().project() {
-                    StateProj::Resolve(fut, stack) => {
-                        let ip = ready!(fut.poll(cx).map_err(Error::Dns))?;
-                        let (svc, config) = stack.take().unwrap();
-                        let addr = SocketAddr::from((ip, config.addr.port()));
-                        this.state
-                            .as_mut()
-                            .set(State::NotReady(svc, Some((addr, config))));
+        let port = address.addr.port().clone();
+        let server_name = address.identity.clone();
+        try_stream! {
+            match address.clone().addr {
+                Addr::Socket(sa) => {
+                    // this should yield the socket address once
+                    // and keep on yielding NotReady forever.
+                    let mut stream = futures::stream::once(future::ok::<Update<Target>, Error>(
+                        Update::Add(vec![(sa, Target::new(sa, address.identity))]),
+                    ))
+                    .chain(tokio::stream::pending());
+
+                    while let Some(value) = stream.next().await {
+                        debug!(?value, "resolved once");
+                        yield value.expect("no errors in this stream");
                     }
-                    StateProj::NotReady(svc, cfg) => {
-                        ready!(svc.poll_ready(cx).map_err(Error::Inner))?;
-                        let (addr, config) = cfg.take().unwrap();
-                        let state = State::make_inner(addr, &config, svc);
-                        this.state.as_mut().set(state);
+                }
+                Addr::Name(name_addr) => {
+                    let mut current: Vec<Target> = Vec::new();
+                    loop {
+                        let mut resolve_ips = Some(dns.resolve_ips(name_addr.name()));
+                        if let Some(resolve_result) = resolve_ips {
+                            match resolve_result.await {
```

Comment:
I think this can be simplified my just matching the `resolve_ips` result

```suggestion
                        match dns.resolve_ips(name_addr.name()).await {
```